sdr_airspy.c

```diff
@@ -0,0 +1,467 @@
+//
+//   All rights reserved.
+//
+//   Redistribution and use in source and binary forms, with or without
+//   modification, are permitted provided that the following conditions are
+//   met:
+//
+//    *  Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//
+//    *  Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//
+//   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+//   "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+//   LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+//   A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+//   HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+//   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+//   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+//   DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+//   THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+//   (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+//   OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "dump1090.h"
+#include "sdr_airspy.h"
+
+#include <airspy.h>
+#include <soxr.h>
+#include <inttypes.h>
+
+static struct {
+    struct airspy_device *dev;
+    uint64_t serial_number;
+    bool enable_lna_agc;
+    bool enable_mixer_agc;
+    bool enable_linearity;
+    bool enable_sensitivity;
+    bool enable_airspy_biast;
+    uint8_t linearity_gain;
+    uint8_t sensitivity_gain;
+    uint8_t mixer_gain;
+    uint8_t lna_gain;
+    uint8_t vga_gain;
+    soxr_t resampler;
+    char *airspy_bytes;
+    char *airspy_scratch;
+    iq_convert_fn converter;
+    struct converter_state *converter_state;
+} AIRSPY;
+
+//
+// =============================== AIRSPY handling ==========================
+//
+
+void airspyInitConfig()
+{
+    AIRSPY.dev = NULL;
+    AIRSPY.serial_number = 0;
+    AIRSPY.enable_lna_agc = false;
+    AIRSPY.enable_mixer_agc = false;
+    AIRSPY.enable_linearity = false;
+    AIRSPY.enable_sensitivity = false;
+    AIRSPY.enable_airspy_biast = false;
+    AIRSPY.linearity_gain = 0;
+    AIRSPY.sensitivity_gain = 0; 
+    AIRSPY.mixer_gain = 8;
+    AIRSPY.lna_gain = 13;
+    AIRSPY.vga_gain = 5;
+    AIRSPY.resampler = NULL; 
+    AIRSPY.converter = NULL;
+    AIRSPY.converter_state = NULL;
+}
+
+static bool parse_uintt(char *s, uint64_t *const value, int utype)
+{
+    uint_fast8_t base = 10;
+    char *s_end;
+    uint64_t u64_val;
+
+    if (s[0] == '-')  // likely forgot a parameter
+        return false;
+
+    if (!isdigit(s[0]))  // we expect at least one number
```

Comment:
strtoull will return s_end == s if there are no digits, so this is already handled below